Tests/DatadogTests/Datadog/Mocks/CoreMocks.swift

```diff
@@ -610,6 +609,73 @@ class NoOpFileReader: SyncReader {
     func markAllFilesAsReadable() {}
 }
 
+/// `AsyncWriter` stub which records events in-memory and allows reading their data back.
+internal class InMemoryWriter: AsyncWriter {
+    let queue = DispatchQueue(label: "com.datadoghq.InMemoryWriter-\(UUID().uuidString)")
+
+    private let encoder: JSONEncoder = .default()
+    private var events: [Data] = []
+
+    func write<T>(value: T) where T: Encodable {
+        queue.async {
+            do {
+                let eventData = try self.encoder.encode(value)
+                self.events.append(eventData)
+                self.waitAndReturnEventsDataExpectation?.fulfill()
+            } catch {
+                assertionFailure("Failed to encode event: \(value)")
+            }
+        }
+    }
+
+    func flushAndCancelSynchronously() {
+        queue.sync {}
+    }
+
+    // MARK: - Retrieving Recorded Data
+
+    private var waitAndReturnEventsDataExpectation: XCTestExpectation?
+
+    /// Waits until given number of events is written and returns data for these events.
+    /// Passing no `timeout` will result with picking the recommended timeout for unit tests.
+    func waitAndReturnEventsData(count: UInt, timeout: TimeInterval? = nil, file: StaticString = #file, line: UInt = #line) -> [Data] {
```

Comment:
This is copied from [`DataUploadWorkerMock`](https://github.com/DataDog/dd-sdk-ios/blob/e2a961ddc9439e523cf234829bd9a27a4d79451d/Tests/DatadogTests/Datadog/Mocks/DataUploadWorkerMock.swift#L82-L119) as `InMemoryWriter` is aimed to replace it entirely after Tracing and RUM are migrated in next PRs.